crates/spin-wasip3-http/src/lib.rs

```diff
@@ -0,0 +1,404 @@
+/// # Variants
+///
+/// - [`Error::ErrorCode`]: Wraps a low-level [`wasip3::http::types::ErrorCode`]
+///   reported by the WASI HTTP runtime (e.g. connection errors, protocol errors).
+///
+/// - [`Error::Other`]: Represents an arbitrary dynamic error implementing
+///   [`std::error::Error`]. This allows integration with external libraries or
+///   application-specific failure types.
+///
+/// - [`Error::Response`]: Contains a full [`wasip3::http::types::Response`]
+///   representing an HTTP-level error (for example, a `4xx` or `5xx` response
+///   that should be treated as an error condition).
+///
+/// # See also
+/// - [`wasip3::http::types::ErrorCode`]: Standard WASI HTTP error codes.
+/// - [`wasip3::http::types::Response`]: Used when an error represents an HTTP response body.
+#[derive(Debug)]
+pub enum Error {
+    /// A low-level WASI HTTP error code.
+    ///
+    /// Wraps [`wasip3::http::types::ErrorCode`] to represent
+    /// transport-level or protocol-level failures.
+    ErrorCode(wasip3::http::types::ErrorCode),
+    /// An error originating from the [`http`] crate.
+    ///
+    /// Covers errors encountered during the construction,
+    /// parsing, or validation of HTTP types (e.g. invalid headers,
+    /// malformed URIs, or protocol violations) within the
+    /// standard Rust [`http`] ecosystem.
+    HttpError(http::Error),
+    /// A dynamic application or library error.
+    ///
+    /// Used for any runtime error that implements [`std::error::Error`],
+    /// allowing flexibility for different error sources.
+    Other(Box<dyn std::error::Error + Send + Sync>),
+    /// An HTTP response treated as an error.
+    ///
+    /// Contains a full [`wasip3::http::types::Response`], such as
+    /// a `404 Not Found` or `500 Internal Server Error`, when
+    /// the response itself represents an application-level failure.
+    Response(wasip3::http::types::Response),
+}
+
+impl std::fmt::Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Error::ErrorCode(e) => write!(f, "{e}"),
+            Error::HttpError(e) => write!(f, "{e}"),
+            Error::Other(e) => write!(f, "{e}"),
+            Error::Response(resp) => match http::StatusCode::from_u16(resp.get_status_code()) {
+                Ok(status) => write!(f, "{status}"),
+                Err(_) => write!(f, "invalid status code {}", resp.get_status_code()),
+            },
+        }
+    }
+}
+
+impl std::error::Error for Error {}
+
+impl From<http::Error> for Error {
+    fn from(err: http::Error) -> Error {
+        match err {
+            err if err.is::<http::method::InvalidMethod>() => {
+                Error::ErrorCode(types::ErrorCode::HttpRequestMethodInvalid)
+            }
+            err if err.is::<http::uri::InvalidUri>() => {
+                Error::ErrorCode(types::ErrorCode::HttpRequestUriInvalid)
+            }
+            err => Error::HttpError(err),
+        }
```

Comment:
I'd suggest leaving it alone here and doing this conversion later if necessary. 🤷 
 
```suggestion
        Error::HttpError(err)
```